File: examples/pull_replicate.py
#!/usr/bin/env python3
from typing import List, Dict, Any, Optional, Callable
from dataclasses import dataclass, astuple
import base64
import functools
import inspect
import logging
import requests
import sqlite3
import sys
import time
import urllib.parse


class LoggingTimer:
    def __init__(self, name):
        self.name = name
        self.s = None

    def __enter__(self):
        self.s = time.time()

    def __exit__(self, a, b, c):
        e = time.time()
        d = e - self.s
        logging.info("  %s took %ss", self.name, f"{d:.3f}")


def trace_timing(fspec: str) -> Callable:
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapped(*args, **kwargs):
            with LoggingTimer(
                fspec.format(**inspect.getcallargs(func, *args, **kwargs))
            ):
                return func(*args, **kwargs)

        return wrapped

    return decorator


@dataclass
class CompressedWeb:
    wid: int
    created: str  # TODO datetime?
    url: str
    status: int
    response: bytes

    @staticmethod
    def from_json(j: Dict[str, Any]) -> "CompressedWeb":
        return CompressedWeb(
            j["id"],
            j["created"],
            j["url"],
            j["status"],
            base64.b64decode(j["response"]),
        )


class Client:
    def __init__(self, username: str, password: str) -> None:
        self.session = requests.Session()
        self.session.auth = (username, password)
        self.session.headers.update(
            {"User-Agent": f"skitter-ro-client-py/0.0.1 +{username}"}
        )
        self.base_url = "https://zst1uv23.fanfic.dev/"

    @trace_timing("fetch_max_wid")
    def fetch_max_wid(self) -> int:
        res = self.session.get(urllib.parse.urljoin(self.base_url, "v0/web/stat"))
        res.raise_for_status()
        j = res.json()
        return int(j["max_wid"])

    @trace_timing(
        "fetch_range_compressed(max_wid={max_wid}, min_wid={min_wid}, url_like={url_like})"
    )
    def fetch_range_compressed(
        self,
        min_wid: int,
        max_wid: int,
        url_like: Optional[str],
    ) -> List[CompressedWeb]:
        res = self.session.get(
            urllib.parse.urljoin(self.base_url, "v0/web/range"),
            params=[
                ("min_wid", min_wid),
                ("max_wid", max_wid),
                ("url_like", url_like),
            ],
        )
        res.raise_for_status()
        j = res.json()
        return [CompressedWeb.from_json(e) for e in j["entries"]]


def init_logging() -> None:
    logging.basicConfig(
        format="%(asctime)s\t%(levelname)s\t%(message)s",
        level=logging.DEBUG,
    )


def get_pool(db_url: str) -> sqlite3.Connection:
    conn = sqlite3.connect(db_url)
    # TODO: log statements?

    conn.execute(open("./sql/001_init.sql", "r").read())
    conn.commit()

    return conn


@trace_timing("pull(url_like={url_like})")
def pull(url_like: str, client: Client, pool: sqlite3.Connection) -> None:
    max_wid = client.fetch_max_wid()
    logging.info("fetched max_wid: %s", max_wid)
    max_wid += 1

    stored_max_wid = pool.execute("select max(id) from web").fetchone()[0]
    stored_max_wid = 0 if stored_max_wid is None else stored_max_wid
    stored_max_wid = max(149470000 - 1, stored_max_wid)

    for next_wid in range(stored_max_wid + 1, max_wid, 1000):
        target_max_wid = min(next_wid + 1000, max_wid)
        pull_block(next_wid, target_max_wid, url_like, client, pool)


@trace_timing("pull_block(url_like={url_like}, min_wid={min_wid}, max_wid={max_wid})")
def pull_block(
    min_wid: int,
    max_wid: int,
    url_like: str,
    client: Client,
    pool: sqlite3.Connection,
) -> None:
    res = client.fetch_range_compressed(
        min_wid,
        max_wid,
        url_like,
    )
    logging.info(
        "fetched block: block_span=%s count=%s",
        max_wid - min_wid,
        len(res),
    )

    pool.executemany(
        "insert into web(id, created, url, status, response) values(?, ?, ?, ?, ?)",
        [astuple(r) for r in res],
    )
    pool.commit()


def main(args: List[str]) -> int:
    init_logging()

    if len(args) != 4:
        print(f"usage: {args[0]} <user> <pass> <url_like>")
        return 1

    username = args[1]
    password = args[2]
    url_like = args[3]

    client = Client(username, password)

    db_url = "./py_web.db"
    pool = get_pool(db_url)

    while True:
        try:
            pull(url_like, client, pool)
        except KeyboardInterrupt:
            break
        time.sleep(60 - time.time() % 60)

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: examples/pull_replicate.rs
use skitter_ro_client::{Client, Url};
use sqlx::sqlite::SqliteConnectOptions;
use sqlx::ConnectOptions;
use sqlx::{sqlite::SqlitePool, sqlite::SqlitePoolOptions};
use std::cmp::{max, min};
use std::str::FromStr;
use tap::prelude::*;
use tokio::time::Duration;
use tracing_log::LogTracer;
use tracing_subscriber::fmt::format::FmtSpan;
use tracing_subscriber::{prelude::*, EnvFilter};

struct Tracer;

impl Tracer {
    pub fn new() -> Self {
        LogTracer::init().expect("failed to set logger");

        let env_filter =
            EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info"));

        tracing::subscriber::set_global_default(
            tracing_subscriber::registry().with(env_filter).with(
                tracing_subscriber::fmt::Layer::new()
                    .with_file(true)
                    .with_line_number(true)
                    .with_span_events(FmtSpan::CLOSE),
            ),
        )
        .expect("failed to set tracing subscriber");

        Self {}
    }
}

impl Drop for Tracer {
    fn drop(&mut self) {}
}

#[tracing::instrument]
async fn get_pool(db_url: &str) -> SqlitePool {
    let conn_opt = SqliteConnectOptions::from_str(db_url)
        .expect("failed to build conn_opt")
        .create_if_missing(true)
        .tap_mut(|c| {
            c.log_statements(tracing_log::log::LevelFilter::Debug);
        });
    let pool = SqlitePoolOptions::new()
        .max_connections(1)
        .connect_with(conn_opt)
        .await
        .expect("failed to connect to sqlite db");

    sqlx::query(&std::fs::read_to_string("./sql/001_init.sql").expect("failed to read schema"))
        .execute(&pool)
        .await
        .expect("failed to create table");

    pool
}

#[tokio::main(flavor = "multi_thread", worker_threads = 1)]
async fn main() {
    let _tracer = Tracer::new();

    let args = std::env::args().collect::<Vec<String>>();
    if args.len() != 4 {
        eprintln!("usage: {} <user> <pass> <url_like>", args[0]);
        return;
    }

    let user = &args[1];
    let pass = &args[2];
    let url_like = &args[3];

    let client = reqwest::Client::new();
    let client = Client::new(
        client,
        Url::parse("https://zst1uv23.fanfic.dev/").unwrap(),
        user,
        pass,
    );

    let db_url = "./web.db";
    let pool = get_pool(db_url).await;

    let mut interval = tokio::time::interval(Duration::from_secs(60));
    interval.tick().await; // First tick completes near instantly.

    loop {
        pull(url_like, &client, &pool).await;
        interval.tick().await;
    }
}

#[tracing::instrument(skip(client, pool))]
async fn pull(url_like: &str, client: &Client<'_>, pool: &SqlitePool) {
    let max_wid = client
        .fetch_stat()
        .await
        .expect("failed to fetch_stat")
        .max_wid;
    tracing::info!(max_wid, "fetched max_wid");
    let max_wid = max_wid + 1; // Cover max_wid with half-open range.

    let stored_max_wid = max(
        149470000 - 1,
        sqlx::query_scalar("select max(id) from web")
            .fetch_one(pool)
            .await
            .unwrap_or(0),
    );

    const BLOCK_SIZE: usize = 1000;
    for next_wid in (stored_max_wid + 1..max_wid).step_by(BLOCK_SIZE) {
        let target_max_wid = min(next_wid + 1000, max_wid);
        pull_block(next_wid, target_max_wid, url_like, client, pool).await;
    }
}

#[tracing::instrument(skip(client, pool))]
async fn pull_block(
    min_wid: i64,
    max_wid: i64,
    url_like: &str,
    client: &Client<'_>,
    pool: &SqlitePool,
) {
    let res = client
        .fetch_range_compressed(min_wid, max_wid, Some(url_like))
        .await
        .expect("failed to fetch_range");
    tracing::info!(
        block_span = max_wid - min_wid,
        count = res.len(),
        "fetched block"
    );

    for r in res.into_iter() {
        sqlx::query("insert into web(id, created, url, status, response) values(?, ?, ?, ?, ?)")
            .bind(r.id)
            .bind(r.created)
            .bind(r.url)
            .bind(r.status)
            .bind(r.response)
            .execute(pool)
            .await
            .expect("failed to insert");
    }
}
